src/gpuarray_reduction.c

```diff
@@ -6,841 +6,4095 @@
+		default:                     if (name && n>0){*name = '\0';} return GA_INVALID_ERROR;
+	}
+}
+GPUARRAY_PUBLIC int   GpuReductionAttr_issensitive  (const GpuReductionAttr*    grAttr){
+	/**
+	 * @brief Returns whether the reduction is "sensitive".
+	 * 
+	 * A reduction is sensitive when its output satisfies at least one of the
+	 * following conditions:
+	 * 
+	 *   - It depends on the exact order of axes in the reduxList
+	 *   - It depends on exact signs of the strides of axes in the reduxList
+	 * 
+	 * Such sensitivity may prevent a flattening of contiguous axes even when it
+	 * would have been otherwise permitted.
+	 * 
+	 * For instance, ARGMIN/ARGMAX have this sensitivity, because the dstArg
+	 * tensor's contents are flattened coordinates into the source tensor, and
+	 * the flattening order is precisely reduxList. Permuting it would thus produce
+	 * incorrect output. Moreover, if the strides of a reduction axis were to be
+	 * reversed for the purpose of flattening the axis into another, the computed
+	 * coordinate would again be incorrect.
+	 * 
+	 * 
+	 * TL;DR: Reduction is sensitive if
+	 *   reduce(x, axis=axisList) != reduce(x, axis=axisList[::-1])
+	 * or
+	 *   reduce(x) != reduce(x[::-1])
+	 * .
+	 */
+	
+	switch (grAttr->op){
+		case GA_REDUCE_MINANDARGMIN:
+		case GA_REDUCE_MAXANDARGMAX:
+		case GA_REDUCE_ARGMIN:
+		case GA_REDUCE_ARGMAX:
+		  return 1;
+		default:
+		  return 0;
+	}
+}
+GPUARRAY_PUBLIC int   GpuReductionAttr_requiresS0   (const GpuReductionAttr*    grAttr){
+	switch (grAttr->op){
+		default: return 1;
+	}
+}
+GPUARRAY_PUBLIC int   GpuReductionAttr_requiresD0   (const GpuReductionAttr*    grAttr){
+	switch (grAttr->op){
+		case GA_REDUCE_ARGMIN:
+		case GA_REDUCE_ARGMAX:
+		  return 0;
+		default:
+		  return 1;
+	}
+}
+GPUARRAY_PUBLIC int   GpuReductionAttr_requiresD1   (const GpuReductionAttr*    grAttr){
+	switch (grAttr->op){
+		case GA_REDUCE_MINANDARGMIN:
+		case GA_REDUCE_MAXANDARGMAX:
+		case GA_REDUCE_ARGMIN:
+		case GA_REDUCE_ARGMAX:
+		  return 1;
+		default:
+		  return 0;
+	}
+}
+GPUARRAY_PUBLIC void  GpuReductionAttr_free         (GpuReductionAttr*          grAttr){
+	free(grAttr);
+}
+GPUARRAY_PUBLIC int   GpuReduction_new              (GpuReduction**             gr,
+                                                     const GpuReductionAttr*    grAttr){
+	if (!gr){
+		return GA_INVALID_ERROR;
+	}
+	if (!grAttr){
+		*gr = NULL;
+		return GA_INVALID_ERROR;
+	}
+	
+	*gr = calloc(1, sizeof(**gr));
+	if (*gr){
+		(*gr)->grAttr = *grAttr;
+		(*gr)->gpuCtx = grAttr->gpuCtx;
+		(*gr)->op     = grAttr->op;
+		(*gr)->nds    = (int)grAttr->maxSrcDims;
+		(*gr)->ndd    = (int)grAttr->maxDstDims;
+		(*gr)->ndr    = (int)(grAttr->maxSrcDims-grAttr->maxDstDims);
+		
+		return reduxGenInit(*gr);
+	}else{
+		return GA_MEMORY_ERROR;
+	}
+}
+GPUARRAY_PUBLIC void  GpuReduction_free             (GpuReduction*              gr){
+	reduxGenCleanup(gr, !GA_NO_ERROR);
```

Comment:
This was an idiotic attempt at freeing the `GpuReduction*` by invoking its cleanup code with an error code, thus pretending its initialization had failed. `reduxGenCleanup()` frees everything but but only if `ret != GA_NO_ERROR`.

Admittedly this could be done better.